appserver/ejb/ejb-http-remoting/client/src/main/java/fish/payara/ejb/http/client/adapter/ClientAdapterCustomizer.java

```diff
@@ -0,0 +1,165 @@
+ * <pre>
+ * {@code
+ * import static fish.payara.ejb.http.client.adapter.ClientAdapterCustomizer.*;
+ *
+ * ...
+ * CompositeClientAdapter.builder()
+ *     .register(customize(JmsStubAdapter.class).matchPrefix("java:comp/jms"),
+ *               customize(new ConnectionFactoryStub()).matchName(Pattern.compile("jms/.+Factory")::matches))
+ *     .build()
+ * }
+ * </pre>
+ */
+public final class ClientAdapterCustomizer implements ClientAdapter {
+    private final Supplier<ClientAdapter> downstreamAdapter;
+    private Predicate<String> namePredicate;
+    private Function<String, String> nameTransformation = Function.identity();
+
+    ClientAdapterCustomizer(Supplier<ClientAdapter> downstreamAdapter) {
+        this.downstreamAdapter = downstreamAdapter;
+    }
+
+    private ClientAdapterCustomizer(ClientAdapterCustomizer previous) {
+        this.downstreamAdapter = previous.downstreamAdapter;
+        this.namePredicate = previous.namePredicate;
+        this.nameTransformation = previous.nameTransformation;
+    }
+
+    /**
+     * Call downstream adapter when name matches, with jndiName parameter transformed.
+     * @param jndiName jndi name requested for lookup
+     * @param remoteContext naming context for remote EJB invocation
+     * @return looked up adapter if downstream adapter returns one
+     * @throws NamingException when downstream adapter throws one
+     */
+    @Override
+    public Optional<Object> makeLocalProxy(String jndiName, Context remoteContext) throws NamingException {
+        if (namePredicate == null || namePredicate.test(jndiName)) {
+            return downstreamAdapter.get().makeLocalProxy(nameTransformation.apply(jndiName), remoteContext);
+        } else {
+            return Optional.empty();
+        }
+    }
+
+
+    /**
+     * Return new customizer with specified predicate for matching a name.
+     * The predicates is replaced rather than composed, any current predicate is replaced with the new one in returned instance.
+     * @param name predicate for matching name
+     * @return new customizer instance
+     */
+    public ClientAdapterCustomizer matchName(Predicate<String> name) {
+        ClientAdapterCustomizer conditional = new ClientAdapterCustomizer(this);
+        conditional.namePredicate = name;
+        return conditional;
+    }
+
+    /**
+     * Return new customizer with specified name transformation function.
+     * The function is replaced rather than composed, any current transformation is replaced with new one in returned instance.
+     * Transformation is applied after predicate specified with {@link #matchName(Predicate)} passes
+     * @param nameTransformation JNDI name transformation
+     * @return new customizer instance
+     */
+    public ClientAdapterCustomizer transformName(Function<String, String> nameTransformation) {
+        ClientAdapterCustomizer conditional = new ClientAdapterCustomizer(this);
```

Comment:
`null` can be set but is not handled later as default of identity is assumed.

Reply:
Ok, will use identity when `null is provided`